### ISSUE TITLE
bash: 61 - : syntax error: operand expected (error token is "- ")

### ISSUE THREAD

**Issue by ammarooo** [akinomyoga/ble.sh#176]
**ble version**:v0.4.0-devel2 and latest commit from master
**Bash version**: 4.1.2(1)-release (x86_64-redhat-linux-gnu) and version 5.0.17(1)-release (x86_64-pc-linux-gnu)

on sourcing of the ble.sh and whenever a command has failed this syntax error appears
![image](https://user-images.githubusercontent.com/7798200/153838664-3b1bdbce-3933-49c2-abe3-75578a1a03e2.png)

-->

**Comment by akinomyoga** [comment]
Thank you for the report! I currently cannot reproduce the error in my environment. Maybe this is caused by interference with your existing configuration. I would like to solve the possible interference caused by `ble.sh`, so could you help me with identifying the problem? I'm going to ask several questions in the following replies.

- **Q1**: Does the problem reproduce with the plain Bash + ble.sh?

  ```bash
  $ PS1='$ ' INPUTRC=/dev/null bash --norc --noprofile      # <- this starts a plain Bash session
  $ source ble-0.4.0-devel2/ble.sh
  $ false      # <- some failing command
        # <- Does the error message reproduced here?
  ```

- **Q2**: Similarly, can you reproduce the error message with the following configuration?

  ```bash
  $ PS1='$ ' bash --norc      # <- this starts a Bash session without your .bashrc
  $ source ble-0.4.0-devel2/ble.sh
  $ false      # <- some failing command
        # <- Does the error message reproduced here?
  ```

- **Q3**: If the error message doesn't reproduce in Q2, can you identify what in your `.bashrc` causes the interference? If you don't know how to identify it, and you are fine with sharing your `.bashrc`, maybe you can instead paste your `.bashrc` (in that case, please replace your personal information that you don't want to share such as the username, server name, etc., if any).

**Comment by ammarooo** [comment]
thanks for your response.
I tried to track this problem via checking the PS1 configuration on my machine.
I think the problem appears when there is a trap to calculate the execution time  and this time is printed to PS1

when I removed this trap from PS1 the error does not appear.
I have something similar to this https://jakemccrary.com/blog/2015/05/03/put-the-last-commands-run-time-in-your-bash-prompt/ in my bashrc.

**Comment by akinomyoga** [comment]
OK! Thank you for the information! I could reproduce it locally! I'll fix it at my side later.

By the way, the latest version of `ble.sh` (but not 0.4.0-devel2) has a similar feature with more resolution. Could you use the following setting until I fix it?

```bash
# Example for ~/.bashrc
source /path/to/ble.sh --attach=none # Please replace "/path/to" with your configuration.

if [[ ${BLE_VERSION-} ]]; then
  # When ble.sh is loaded, use the built-in feature of ble.sh
  ble-import contrib/prompt-elapsed
  _ps1_time='\q{contrib/elapsed}'
else
  function timer_start {
    echo "DEBUG:($1)" >/dev/tty
    timer=${timer:-$SECONDS}
  }

  function timer_stop {
    timer_show=$(($SECONDS - $timer))
    unset timer
  }

  trap 'timer_start "$BASH_COMMAND"' DEBUG

  if [ "$PROMPT_COMMAND" == "" ]; then
    PROMPT_COMMAND="timer_stop"
  else
    PROMPT_COMMAND="$PROMPT_COMMAND; timer_stop"
  fi

  _ps1_time='${timer_show}s'
fi

PS1='[last: '$_ps1_time'][\w]$ '

[[ ${BLE_VERSION-} ]] && ble-attach
```

If you would like to have more detailed timing information, you may also try `contrib/config/execmark`:

```bash
# ~/.blerc

# Enabled the elapsed-time line
ble-import -d contrib/config/execmark

# Condition to show the elapsed-time line. Unit is millisecond.
# If you want to show it always, you can just specify exec_elapsed_enabled=1.
bleopt exec_elapsed_enabled='sys+usr>=1000'
```

**Comment by ammarooo** [comment]
Thanks, contrib/prompt-elapsed looks like what I need..
I think there are no further steps to be done. We can close this ticket.

**Comment by akinomyoga** [comment]
Yeah, thank you for the further investigation! Actually, I had already identified the causes when I replied previously, and also the phenomena you have reported are what I have already expected, tested, and confirmed. These phenomena are caused by the following design of `ble.sh`:

1. `ble.sh` also uses `DEBUG` trap internally for some operations. With the current version, the original `DEBUG` trap before sourcing `ble.sh` is lost. I think it is not difficult to save the original `DEBUG`, so I will later add the code to save the original `DEBUG` trap.
2. Another incompatibility is that `ble.sh` intentionally disables the `DEBUG` trap for its internal processing and only enables the `DEBUG` trap for the user command (executed by <kbd>RET</kbd>, etc.). The internal processing includes `PROMPT_COMMAND`, so the `DEBUG` trap will not be called for the commands in `PROMPT_COMMAND` with `ble.sh`. This is different from the plain Bash where the `DEBUG` trap is called also for the commands in `PROMPT_COMMAND`. Maybe I should enable the `DEBUG` trap when `ble.sh` calls `PROMPT_COMMAND`. I'll later think about whether there is a better solution.

> after some debugging..
> `timer_show=$((SECONDS - timer))` instead of `timer_show=$(($SECONDS - $timer))`
> solved the issue of the syntax error that is being reported.

Yes, but I think the calculation result would be still wrong. When the error was originally output, the variable `timer` is unset. In that case, `timer` in the arithmetic context will be evaluated as `0`. So `$((SECONDS - timer))` will result in the total time since the shell session started instead of the command execution time of the previous command.

> if the ble.sh is source first and then the PS1 is configured... it is working.

I think it will still fail when the executed commands are something like

```bash
[...]$ <RET>
[...]$ # When the comment is executed.<RET>
[...]$ function func { echo this is a function definition; }<RET>
```

> but if the ble.sh is sourced after setting of the PS1 above the PS1 is not updated correctly

Yes, this is because the original `DEBUG` trap before `ble.sh` is sourced is overwritten by `ble.sh` and lost (as already explained).

**Comment by akinomyoga** [comment]
Sorry for the delay. The fix was trickier than I initially thought. I realized that we cannot get the value of the DEBUG trap from inside of the sourced script (unless in the debug mode of Bash). There is still some limitation, but I mostly completed the fix locally. I'm now testing the behavior in various setups and fixing small quirks locally. Hope I can push the fix to GitHub this night.

**Comment by akinomyoga** [comment]
I have pushed a fix ec2a67a to the `master` branch for the treatment of the `DEBUG` trap. However, there are still some limitations on the `DEBUG` behavior on the startup (i.e., small behavioral difference between the plain Bash on the startup). Once the session starts, there should be no problems unless one needs to use `FUNCNAME`, `BASH_SOURCE`, `BASH_LINENO`, etc. in the `DEBUG` trap, or to dynamically rewrite the `DEBUG` trap.

<details><summary>Details: Behavior difference of the DEBUG trap on the startup</summary>

- If `ble.sh` is loaded first (as recommended in README Sec. 1.3), the `DEBUG` trap set by users and other configurations should work mostly as expected.
  - However, when all of the following conditions are satisfied:
    - the Bash version is 5.0 or below,
    - `ble.sh` uses `--attach=prompt` (default when no `--attach` is specified) for attaching the session,
    - and the user or other configuration appends settings to `PROMPT_COMMAND` prepared by `ble.sh`,

    the `DEBUG` trap set up by the user or other configuration (if any) is disabled while executing the appended settings for the first time. The problem happens only for the first evaluation of `PROMPT_COMMAND`; The `DEBUG` trap will be enabled for the second time and later.
- When `ble.sh` is loaded after the user or another configuration set up the `DEBUG` trap, when one of the following conditions is satisfied, the `DEBUG` trap will be disabled until the user inputs something from the keyboard or the terminal sends other internal data.
  - When the Bash version is 3.1 or lower.
  - When `ble-attach` is called from inside another function or from a file that is not the rcfile.
  - When the option `--attach=prompt` is specified or other `--attach=*` / `--noattach` is not specified to `source ble.sh`, and the user or another configuration rewrites `PROMPT_COMMAND` set by `ble.sh` so that `ble.sh` is not called from the top level of `PROMPT_COMMAND`.
  - When the option `--attach=attach` is specified to `source ble.sh`, and *NOT(* `set -T` is specified, and the file name of the Bash initialization file (rcfile) is one of `.bashrc`, `.bash_profile`, or `.profile`, and `ble.sh` is directly loaded from the rcfile *)*, and *NOT(* `set -T` is specified, and `source ble.sh` is directly executed from the command line *)*.
 
</details>

Because of these small differences, the setting in "[*Put the last command's run time in your Bash prompt - Jake McCrary*](https://jakemccrary.com/blog/2015/05/03/put-the-last-commands-run-time-in-your-bash-prompt/)" still produces or behaves strangely on the `ble.sh` startup when one of the above conditions is satisfied.
- Nevertheless. the problem only occurs on the startup.
- When one uses Bash 5.1 or above, and the `DEBUG` trap is set after `ble.sh` is loaded, the problem doesn't happen.
- When `$(($SECONDS - $timer))` is rewritten to `$((SECONDS - timer))`, the error message will not appear.

I actually think the setting "[*Put the last command's run time in your Bash prompt - Jake McCrary*](https://jakemccrary.com/blog/2015/05/03/put-the-last-commands-run-time-in-your-bash-prompt/)" is not carefully designed so that it works with other settings. It can be easily broken e.g. when another configuration appends a setting to `PROMPT_COMMAND` by `PROMPT_COMMAND+=$PROMPT_COMMAND$'\n'NEW_SETTING_BY_OTHER_CONFIG`, etc. where the resulting timing information becomes wrong. Such a configuration includes `bash-preexec.sh` and other shell-terminal ingtegration scripts. Relying on `DEBUG` is also in general not so stable, and it may also introduce a large overhead.

**Comment by akinomyoga** [comment]
Thank you!
